### PR TITLE
Type suggestion API performance improvements

### LINKED ISSUE #1762: Improve `type_suggestions` API performance
## Problem
`/type_suggestions/` API used when importing and previewing a datafile takes really long(around 15 secs). We should reduce the response time to increase UX

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #1762

Adds `cached_metadata` and removes some duplicate queries to improve performance.

**Technical Details**

The performance improvements made by the PR are as follows
- SqlAlchemy queries were reduced from 6865 to 1173 queries

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/columns/operations/alter.py [hunk 1]
```diff
@@ -182,22 +170,21 @@ def change_column_nullable(table_oid, column_attnum, engine, connection, nullabl
...
+    metadata = metadata if metadata else get_empty_metadata()
     table = reflect_table_from_oid(
         table_oid,
         engine,
         connection_to_use=connection,
-        # TODO reuse metadata
-        metadata=get_empty_metadata(),
+        metadata=metadata,
     )
     # TODO reuse metadata
```
Comment (mathemancer) [inline db/columns/operations/alter.py:181]: This comment is obsolete, no?

File: db/schemas/operations/select.py [hunk 2]
```diff
@@ -10,14 +10,14 @@
...
 
 
-def reflect_schema(engine, name=None, oid=None):
+def reflect_schema(engine, name=None, oid=None, metadata=None):
     # If we have both arguments, the behavior is undefined.
     try:
         assert name is None or oid is None
     except AssertionError as e:
         raise e
     # TODO reuse metadata
```
Comment (mathemancer) [inline db/schemas/operations/select.py:19]: I think this comment can go now.

**Review by mathemancer — APPROVED** [review]
Looks great. I like the systematic method you're using. I also appreciate the extraneous reflection you took out from the `alter_column_type`

I found a couple of spots where there's some leftover comments. Please remove those before merging.

File: db/columns/operations/alter.py (hunk 1 above)
Comment (silentninja) [reply db/columns/operations/alter.py:181]: Thanks for the catch!

**Pull request merged**